File: code-tiny/tools/flutter/cache.py
"""Versioned import dependency cache for incremental Dart analysis."""

from __future__ import annotations

from dataclasses import dataclass, field
import json
from pathlib import Path
from typing import Dict, Iterable, Mapping, Set

from .models import AnalysisFacts, NodeRecord, SummaryRecord


CACHE_VERSION = 1
DEPENDENCY_RELATIONSHIPS = frozenset({"IMPORTS", "EXPORTS", "HAS_PART", "PART_OF"})


@dataclass
class DependencyIndex:
    dependencies: Dict[str, Set[str]] = field(default_factory=dict)

    @classmethod
    def from_facts(cls, facts: AnalysisFacts) -> "DependencyIndex":
        owner = {node.identity: node.evidence.file for node in facts.nodes}
        dependencies: Dict[str, Set[str]] = {}
        for edge in facts.edges:
            if edge.relationship not in DEPENDENCY_RELATIONSHIPS:
                continue
            source = owner.get(edge.source)
            target = owner.get(edge.target)
            if source and target and source != target:
                dependencies.setdefault(source, set()).add(target)
        return cls(dependencies)

    def impacted_files(self, changed: Iterable[str], deleted: Iterable[str] = ()) -> Set[str]:
        impacted = {str(path) for path in changed} | {str(path) for path in deleted}
        reverse: Dict[str, Set[str]] = {}
        for source, targets in self.dependencies.items():
            for target in targets:
                reverse.setdefault(target, set()).add(source)
        queue = list(impacted)
        while queue:
            target = queue.pop()
            for dependent in reverse.get(target, set()):
                if dependent not in impacted:
                    impacted.add(dependent)
                    queue.append(dependent)
        return impacted

    def save(self, path: str | Path) -> None:
        target = Path(path)
        target.parent.mkdir(parents=True, exist_ok=True)
        value = {
            "version": CACHE_VERSION,
            "dependencies": {key: sorted(values) for key, values in sorted(self.dependencies.items())},
        }
        target.write_text(json.dumps(value, indent=2, sort_keys=True) + "\n", encoding="utf-8")

    @classmethod
    def load(cls, path: str | Path) -> "DependencyIndex":
        value = json.loads(Path(path).read_text(encoding="utf-8"))
        if int(value.get("version", -1)) != CACHE_VERSION:
            raise ValueError(
                f"unsupported Dart dependency cache version {value.get('version')!r}; expected {CACHE_VERSION}"
            )
        raw = value.get("dependencies", {})
        if not isinstance(raw, Mapping):
            raise ValueError("Dart dependency cache dependencies must be an object")
        return cls({str(key): {str(item) for item in values} for key, values in raw.items()})


def select_incremental_facts(facts: AnalysisFacts, paths: Iterable[str]) -> AnalysisFacts:
    """Select facts owned by impacted files and retain endpoints as non-writing references."""
    impacted = {str(path) for path in paths}
    if not impacted:
        return AnalysisFacts(
            header=facts.header,
            nodes=(),
            edges=(),
            diagnostics=(),
            summary=SummaryRecord(
                processed_files=0,
                skipped_files=facts.summary.processed_files,
                error_count=0,
                elapsed_ms=facts.summary.elapsed_ms,
            ),
        )
    nodes_by_identity = {node.identity: node for node in facts.nodes}
    owned = {node.identity: node for node in facts.nodes if node.evidence.file in impacted}
    edges = tuple(edge for edge in facts.edges if edge.evidence.file in impacted)
    endpoint_ids = {identity for edge in edges for identity in (edge.source, edge.target)}
    selected = dict(owned)
    for identity in endpoint_ids - owned.keys():
        node = nodes_by_identity.get(identity)
        if node is None:
            continue
        selected[identity] = NodeRecord(
            identity=node.identity,
            kind=node.kind,
            properties={**node.properties, "_reference_only": True},
            evidence=node.evidence,
        )
    diagnostics = tuple(
        item for item in facts.diagnostics if item.evidence is None or item.evidence.file in impacted
    )
    processed = len({node.evidence.file for node in owned.values() if node.kind == "file"})
    return AnalysisFacts(
        header=facts.header,
        nodes=tuple(sorted(selected.values(), key=lambda node: node.identity)),
        edges=edges,
        diagnostics=diagnostics,
        summary=SummaryRecord(
            processed_files=processed,
            skipped_files=max(0, facts.summary.processed_files - processed),
            error_count=sum(item.severity == "error" for item in diagnostics),
            elapsed_ms=facts.summary.elapsed_ms,
        ),
    )

File: rust/crates/analyzer-dart/src/cache.rs
//! Rust port của `tools/flutter/cache.py` — versioned import dependency
//! cache cho incremental Dart analysis (impact expansion + selection).

use std::collections::{BTreeMap, BTreeSet};
use std::path::Path;

use serde_json::Value;

use crate::models::{AnalysisFacts, NodeRecord, SummaryRecord};

pub const CACHE_VERSION: i64 = 1;
const DEPENDENCY_RELATIONSHIPS: [&str; 4] = ["IMPORTS", "EXPORTS", "HAS_PART", "PART_OF"];

/// `DependencyIndex` — file → tập file nó phụ thuộc (qua import edges).
#[derive(Debug, Clone, Default)]
pub struct DependencyIndex {
    pub dependencies: BTreeMap<String, BTreeSet<String>>,
}

impl DependencyIndex {
    /// `from_facts` — owner map identity → evidence.file rồi gom dependency
    /// theo relationship.
    pub fn from_facts(facts: &AnalysisFacts) -> Self {
        let owner: BTreeMap<&str, &str> = facts
            .nodes
            .iter()
            .map(|node| (node.identity.as_str(), node.evidence.file.as_str()))
            .collect();
        let mut dependencies: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
        for edge in &facts.edges {
            if !DEPENDENCY_RELATIONSHIPS.contains(&edge.relationship.as_str()) {
                continue;
            }
            let Some(source) = owner.get(edge.source.as_str()) else {
                continue;
            };
            let Some(target) = owner.get(edge.target.as_str()) else {
                continue;
            };
            if source == target {
                continue;
            }
            dependencies
                .entry((*source).to_string())
                .or_default()
                .insert((*target).to_string());
        }
        Self { dependencies }
    }

    /// `impacted_files` — changed ∪ deleted + closure ngược theo dependency.
    pub fn impacted_files(
        &self,
        changed: &BTreeSet<String>,
        deleted: &BTreeSet<String>,
    ) -> BTreeSet<String> {
        let mut impacted: BTreeSet<String> = changed.union(deleted).cloned().collect();
        let mut reverse: BTreeMap<&str, Vec<&str>> = BTreeMap::new();
        for (source, targets) in &self.dependencies {
            for target in targets {
                reverse.entry(target.as_str()).or_default().push(source);
            }
        }
        let mut queue: Vec<String> = impacted.iter().cloned().collect();
        while let Some(target) = queue.pop() {
            if let Some(dependents) = reverse.get(target.as_str()) {
                for dependent in dependents {
                    if impacted.insert((*dependent).to_string()) {
                        queue.push((*dependent).to_string());
                    }
                }
            }
        }
        impacted
    }

    /// `save` — JSON byte-compatible với bản Python (indent 2, sort keys).
    pub fn save(&self, path: &Path) -> Result<(), String> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
        }
        let dependencies: BTreeMap<&str, Vec<&str>> = self
            .dependencies
            .iter()
            .map(|(key, values)| (key.as_str(), values.iter().map(String::as_str).collect()))
            .collect();
        let mut root = serde_json::Map::new();
        root.insert("version".to_string(), Value::from(CACHE_VERSION));
        root.insert(
            "dependencies".to_string(),
            serde_json::to_value(dependencies).map_err(|e| e.to_string())?,
        );
        let value = sorted_value(&Value::Object(root));
        let mut text = serde_json::to_string_pretty(&value).map_err(|e| e.to_string())?;
        text.push('\n');
        std::fs::write(path, text).map_err(|e| e.to_string())
    }

    /// `load` — version mismatch → Err (caller fallback full re-analysis).
    pub fn load(path: &Path) -> Result<Self, String> {
        let text = std::fs::read_to_string(path).map_err(|e| e.to_string())?;
        let value: Value = serde_json::from_str(&text).map_err(|e| e.to_string())?;
        let version = value.get("version").and_then(Value::as_i64).unwrap_or(-1);
        if version != CACHE_VERSION {
            return Err(format!(
                "unsupported Dart dependency cache version {version:?}; expected {CACHE_VERSION}"
            ));
        }
        let raw = value.get("dependencies").cloned().unwrap_or(Value::Null);
        let Value::Object(map) = raw else {
            return Err("Dart dependency cache dependencies must be an object".to_string());
        };
        let mut dependencies = BTreeMap::new();
        for (key, values) in map {
            let Value::Array(items) = values else {
                continue;
            };
            let set: BTreeSet<String> = items
                .into_iter()
                .filter_map(|item| item.as_str().map(str::to_string))
                .collect();
            dependencies.insert(key, set);
        }
        Ok(Self { dependencies })
    }
}

/// `select_incremental_facts` — facts của file impacted, endpoint ngoài
/// impacted được giữ làm `_reference_only` (không ghi row, chỉ giữ id).
pub fn select_incremental_facts(
    facts: &AnalysisFacts,
    paths: &BTreeSet<String>,
) -> AnalysisFacts {
    if paths.is_empty() {
        return AnalysisFacts {
            header: facts.header.clone(),
            nodes: Vec::new(),
            edges: Vec::new(),
            diagnostics: Vec::new(),
            summary: SummaryRecord {
                record_type: "summary",
                processed_files: 0,
                skipped_files: facts.summary.processed_files,
                error_count: 0,
                elapsed_ms: facts.summary.elapsed_ms,
            },
        };
    }
    let nodes_by_identity: BTreeMap<&str, &NodeRecord> = facts
        .nodes
        .iter()
        .map(|node| (node.identity.as_str(), node))
        .collect();
    let owned: BTreeMap<&str, &NodeRecord> = facts
        .nodes
        .iter()
        .filter(|node| paths.contains(&node.evidence.file))
        .map(|node| (node.identity.as_str(), node))
        .collect();
    let edges: Vec<_> = facts
        .edges
        .iter()
        .filter(|edge| paths.contains(&edge.evidence.file))
        .cloned()
        .collect();
    let mut endpoint_ids: BTreeSet<&str> = BTreeSet::new();
    for edge in &edges {
        endpoint_ids.insert(edge.source.as_str());
        endpoint_ids.insert(edge.target.as_str());
    }
    let mut selected: BTreeMap<String, NodeRecord> = owned
        .iter()
        .map(|(identity, node)| ((*identity).to_string(), (*node).clone()))
        .collect();
    for identity in endpoint_ids {
        if owned.contains_key(identity) {
            continue;
        }
        let Some(node) = nodes_by_identity.get(identity) else {
            continue;
        };
        let mut properties = node.properties.clone();
        properties.insert("_reference_only".to_string(), serde_json::json!(true));
        selected.insert(
            (*identity).to_string(),
            NodeRecord {
                record_type: "node",
                identity: node.identity.clone(),
                kind: node.kind.clone(),
                properties,
                evidence: node.evidence.clone(),
            },
        );
    }
    let diagnostics: Vec<_> = facts
        .diagnostics
        .iter()
        .filter(|item| {
            item.evidence
                .as_ref()
                .map(|ev| paths.contains(&ev.file))
                .unwrap_or(true)
        })
        .cloned()
        .collect();
    let processed = owned
        .values()
        .filter(|node| node.kind == "file")
        .map(|node| node.evidence.file.as_str())
        .collect::<BTreeSet<_>>()
        .len();
    let error_count = diagnostics
        .iter()
        .filter(|item| item.severity == "error")
        .count();
    AnalysisFacts {
        header: facts.header.clone(),
        nodes: selected.into_values().collect(),
        edges,
        diagnostics,
        summary: SummaryRecord {
            record_type: "summary",
            processed_files: processed,
            skipped_files: facts.summary.processed_files.saturating_sub(processed),
            error_count,
            elapsed_ms: facts.summary.elapsed_ms,
        },
    }
}

/// Sắp key object đệ quy — đảm bảo JSON artifact/cache sort_keys giống
/// Python kể cả khi `preserve_order` được bật ở đâu đó trong build graph.
pub fn sorted_value(value: &Value) -> Value {
    match value {
        Value::Object(map) => {
            let keys: BTreeMap<&str, &Value> = map
                .iter()
                .map(|(key, item)| (key.as_str(), item))
                .collect();
            Value::Object(
                keys.into_iter()
                    .map(|(key, item)| (key.to_string(), sorted_value(item)))
                    .collect(),
            )
        }
        Value::Array(items) => Value::Array(items.iter().map(sorted_value).collect()),
        other => other.clone(),
    }
}
